File: test-case/char/chars2d/chars2d.php
<?php
require_once 'Soda/Unittest/index.php';

use Soda\Unittest\TestWork;
use Soda\Unittest\Utils;

class Solution {

    /**
     * @param string[][] $matrix
     * @return string[][]
     */
    function toUpper($matrix) {
        $diff = ord('a') - ord('A');
        for ($i = 0; $i < count($matrix); ++$i) {
            for ($j = 0; $j < count($matrix[0]); ++$j) {
                $matrix[$i][$j] = chr(ord($matrix[$i][$j]) - $diff);
            }
        }
        return $matrix;
    }
}

$work = TestWork::create(new Solution(), 'toUpper', $argv[0]);
// $work->validator = fn($e, $r) => <boolean result>
$work->compareSerial = true;
echo $work->run(Utils::fromStdin());

File: test-case/char/chars2d/Chars2d.scala
import soda.scala.unittest._

import scala.collection.mutable
import scala.reflect.runtime.universe.typeOf

class Solution {}
object Solution {
  def toUpper(matrix: List[List[Char]]): Array[Array[Char]] = {
    val diff = 'a' - 'A'
    val mx = Array.ofDim[Char](matrix.size, matrix(0).size)
    for (i <- matrix.indices) {
      for (j <- matrix(0).indices) {
        mx(i)(j) = (matrix(i)(j) - diff).toChar
      }
    }
    return mx
  }
}

class Chars2d extends (String => String) {
  override def apply(text: String): String = {
    val work = TestWork.forObject(typeOf[Solution], "toUpper")
    // val work = TestWork.forStruct(...)
    // work.setValidator((R, R) => Boolean)
    work.compareSerial = true
    work.run(text)
  }
}

object Chars2d extends App {
  println(new Chars2d()(Utils.fromStdin()))
}
